File: express-backend/src/views/message.js
import fs from "fs/promises";
import cuid from "cuid";
import express from "express";
import { matchedData } from "express-validator";

import * as validator from "../validators/message.js";
import { updateUserForNewMessage } from "../socket.js";
import prisma from "../db.js";
import {
    excludeFieldsFromObject,
    formatFileFields,
    getPrivateFileDirectory,
    paginate,
} from "../helpers.js";

const router = express.Router();

router.post("/", validator.messageValidator, async (req, res) => {
    const areFriends = await prisma.friendship.findFirst({
        where: {
            OR: [
                { requestedById: req.user.id, acceptedById: req.params.id },
                { requestedById: req.params.id, acceptedById: req.user.id },
            ],
            acceptedAt: { not: null },
        },
    });

    if (!areFriends) return res.sendStatus(403);

    const messageData = {
        content: matchedData(req).content,
        senderId: req.user.id,
        recipientId: matchedData(req).recipientId,
    };

    if (!req.file && !messageData.content) return res.sendStatus(400);

    if (req.file) {
        const fileName = `${cuid()}.${req.file.mimetype.split("/")[1]}`;
        fs.writeFile(getPrivateFileDirectory(fileName), req.file.buffer);
        messageData.file = fileName;
    }

    await prisma.message.create({ data: messageData });
    updateUserForNewMessage(messageData.recipientId);
    return res.sendStatus(201);
});

router.get("/chats", async (req, res) => {
    const result = await prisma.$queryRaw`
        WITH
            "ChatTable" AS (
                SELECT
                    "recipientId" AS "friendId",
                    "sentAt",
                    CONCAT (
                        'You: ', 
                        CASE 
                            WHEN "content" IS NULL
                                THEN '📎 Attachment'
                            ELSE 
                                "content"
                        END
                    ) as "content",
                    false as "attention"
                FROM
                    "Message"
                WHERE
                    "senderId" = ${req.user.id}
                UNION
                SELECT
                    "senderId" AS "friendId",
                    "sentAt",
                    CASE 
                        WHEN "content" IS NULL 
                            THEN '📎 Attachment'
                        ELSE
                            "content" 
                    END as "content",
                    NOT "read" as "attention"
                FROM
                    "Message"
                WHERE
                    "recipientId" = ${req.user.id}
            )
        SELECT
            "friendId",
            "firstName",
            "lastName",
            "profileImage",
            "lastActive",
            "sentAt",
            "content",
            "attention"
        FROM
            "ChatTable" INNER JOIN "User" ON "ChatTable"."friendId" = "User"."id"
        WHERE
            ("friendId", "sentAt") IN (
                SELECT
                    "friendId",
                    MAX("sentAt")
                FROM
                    "ChatTable"
                GROUP BY
                    "friendId"
            )
        ORDER BY "sentAt" DESC
    `;

    return res.json(result);
});

router.get("/:id", async (req, res) => {
    await prisma.message.updateMany({
        where: { senderId: req.params.id, recipientId: req.user.id },
        data: { read: true },
    });

    return res.json(
        (
            (await prisma.message.findMany({
                where: {
                    OR: [
                        { senderId: req.user.id, recipientId: req.params.id },
                        { senderId: req.params.id, recipientId: req.user.id },
                    ],
                },
                orderBy: {
                    sentAt: "desc",
                },
                ...paginate(req),
            })) || []
        ).map((m) => formatFileFields(m, ["file"], true))
    );
});

export default router;

File: django-backend/api/views/message.py
from typing import List

from ninja import Router, Form, File, UploadedFile
from ninja.pagination import paginate
from django.db import connection
from django.db.models import Q
from pydantic import UUID4

from api.helpers import create_message, dict_fetchall
from api.models import Friendship, Message
from api.schemas import MessageIn, MessageOut, ChatOut


router = Router(tags=["message"])


@router.post("", response={201: str, 400: str, 403: str})
def send_message(request, data: Form[MessageIn], file: File[UploadedFile] = None):
    friendship = Friendship.objects.filter(
        (Q(requested_by=request.auth, accepted_by=data.recipient_id))
        | (Q(requested_by=data.recipient_id, accepted_by=request.auth)),
        accepted_at__isnull=False,
    ).first()

    if not friendship:
        return 403, "Friendship not found"

    if not data.content and not file:
        return 400, "No content provided"

    create_message(request.auth.id, data.recipient_id, data.content, file)
    return 201, "Message sent"


@router.get("/unread", response=int)
def unread_messages_count(request):
    return Message.objects.filter(recipient_id=request.auth, read=False).count()


@router.get("/chats", response=List[ChatOut])
def chats(request):
    with connection.cursor() as cursor:
        cursor.execute(
            """
                WITH
                    "chat_table" AS (
                        SELECT
                            "recipient_id" AS "friend_id",
                            "sent_at",
                            CONCAT (
                                'You: ', 
                                CASE 
                                    WHEN "content" = ''
                                        THEN '📎 Attachment'
                                    ELSE 
                                        "content"
                                END
                            ) as "content",
                            false as "attention"
                        FROM
                            "api_message"
                        WHERE
                            "sender_id" = %(id)s
                        UNION
                        SELECT
                            "sender_id" AS "friend_id",
                            "sent_at",
                            CASE 
                                WHEN "content" = '' 
                                    THEN '📎 Attachment'
                                ELSE
                                    "content" 
                            END as "content",
                            NOT "read" as "attention"
                        FROM
                            "api_message"
                        WHERE
                            "recipient_id" = %(id)s
                    )
                SELECT
                    "friend_id",
                    "first_name",
                    "last_name",
                    "profile_image",
                    "last_active",
                    "sent_at",
                    "content",
                    "attention"
                FROM
                    "chat_table" INNER JOIN "api_user" ON "chat_table"."friend_id" = "api_user"."id"
                WHERE
                    ("friend_id", "sent_at") IN (
                        SELECT
                            "friend_id",
                            MAX("sent_at")
                        FROM
                            "chat_table"
                        GROUP BY
                            "friend_id"
                    )
                ORDER BY "sent_at" DESC 
            """,
            {"id": request.auth.id},
        )

        return dict_fetchall(cursor)


@router.get("/{id}", response=List[MessageOut])
@paginate
def get_messages(request, id: UUID4):
    Message.objects.filter(recipient_id=request.auth, sender_id=id, read=False).update(
        read=True
    )

    return Message.objects.filter(
        Q(sender_id=request.auth, recipient_id=id)
        | Q(sender_id=id, recipient_id=request.auth)
    ).order_by("-sent_at")
